src/commands/synthetics/run-tests-lib.ts

```diff
@@ -192,52 +168,117 @@ const getTestListBySearchQuery = async (
   }))
 }
 
-export const getTestsList = async (
+export const getFinalVersionOfTestsFromCi = async (
   api: APIHelper,
   config: RunTestsCommandConfig,
   reporter: MainReporter,
-  suites: Suite[] = []
+  suites?: Suite[]
 ) => {
-  // If "testSearchQuery" is provided, always default to running it.
-  if (config.testSearchQuery) {
-    const testsToTriggerBySearchQuery = await getTestListBySearchQuery(api, config.global, config.testSearchQuery)
+  let testConfig: TriggerConfig[]
+  // TODO: Clean up locations as part of SYNTH-12989
+  const configFromEnvironment = config.locations?.length ? {locations: config.locations} : {}
+
+  let testsToTrigger = config.publicIds.map((id) => ({
+    config: {
+      ...config.global,
+      ...configFromEnvironment,
+    },
+    id,
+  }))
 
-    if (testsToTriggerBySearchQuery.length > MAX_TESTS_TO_TRIGGER) {
-      reporter.error(
-        `More than ${MAX_TESTS_TO_TRIGGER} tests returned by search query, only the first ${MAX_TESTS_TO_TRIGGER} will be fetched.\n`
+  if (!testsToTrigger.length && config.testSearchQuery) {
+    try {
+      testsToTrigger = await getTestsFromSearchQuery(api, config, reporter)
+      testsToTrigger = testsToTrigger.map((test) => ({
+        ...test,
+        config: {
+          ...configFromEnvironment,
+          ...test.config,
+        },
+      }))
```

Comment:
This part is a bit confusing to me, especially given the now duplicated overrides (between `config.global` and `configFromEnvironment`, between `configFromEnvironment` and `test.config`, and between `config.global`, `configFromEnvironment` and `test.config`). Could we try to merge these three overrides together?

The following is probably over-simplistic, and missing a lot of edge cases I couldn't foresee, but hopefully it gets the idea across, and we can talk about it.

```typescript
// grab all the test config overrides from different sources
const defaultTestConfigOverride = config.global
const testConfigOverridesFromEnv = config.locations?.length ? {locations: config.locations} : {}
const testConfigOverridesFromTestFiles = getTestConfigOverridesFromTestFiles(config, reporter, suites)

// grab the list of tests to trigger from command line, config file, search query or test config files
const testsToTrigger =
  config.publicIds ??
  config.testSearchQuery && await getTestsFromSearchQuery(api, config, reporter) ??
  testConfigOverridesFromTestFiles.map(({id}) => id)

// Create the overrides required for the list of tests to trigger
const testConfigOverrides = testsToTrigger.map(id => ({
  id,
  config: {
    ...defaultTestConfigOverride,
    ...testConfigOverridesFromEnv,
    ...testConfigOverridesFromTestFiles.find(testConfigOverrideFromTestFile => testConfigOverrideFromTestFile.id === id) ?? {}
  }
})
```